src/block/block_compact.c

```diff
@@ -65,6 +70,295 @@ __wt_block_compact_get_progress_stats(WT_SESSION_IMPL *session, WT_BM *bm,
+     */
+
+    __wt_spin_lock(session, &block->live_lock);
+
+    compact_pct_tenths = block->compact_pct_tenths;
+    extra_space = 0;
+    file_size = block->size;
+    pages_to_move = 0;
+    total_pages_to_move = 0;
+    write_off = 0;
+
+#define WT_EXT_SIZE_TO_LEAF_PAGES(ext_size)                          \
+    (uint64_t)((ext_size) / depth1_size * leaves_per_internal_page + \
+      ((ext_size) % depth1_size) / avg_block_size)
+
+    for (iteration = 0;; iteration++) {
+        compact_start_off = file_size - compact_pct_tenths * file_size / 10;
+        if (write_off >= compact_start_off)
+            break;
+        __wt_verbose_debug2(session, WT_VERB_COMPACT,
+          "%s: estimating -- pass %d: file size: %" PRId64 ", compact offset: %" PRId64
+          ", will move blocks from the last %d%% of the file",
+          block->name, iteration, file_size, compact_start_off, compact_pct_tenths * 10);
+
+        /* Estimate how many pages we would like to move, just using the live checkpoint. */
+        last = compact_start_off;
+        WT_EXT_FOREACH_FROM_OFFSET_INCL(ext, &block->live.avail, compact_start_off)
+        {
+            off = ext->off;
+            size = ext->size;
+            WT_ASSERT(session, off >= compact_start_off || off + size >= compact_start_off);
+
+            __block_compact_trim_extent(session, compact_start_off, file_size, &off, &size);
+            if (off >= compact_start_off && size <= 0)
+                break;
+
+            if (off > last) {
+                n = WT_EXT_SIZE_TO_LEAF_PAGES(off - last);
+                pages_to_move += n;
+                __wt_verbose_debug3(session, WT_VERB_COMPACT,
+                  "%s: estimating -- %" PRIu64 " pages to move between %" PRId64 " and %" PRId64,
+                  block->name, n, last, off);
+            }
+            last = off + size;
+        }
+        n = WT_EXT_SIZE_TO_LEAF_PAGES(file_size - last);
+        pages_to_move += n;
+        __wt_verbose_debug3(session, WT_VERB_COMPACT,
+          "%s: estimating -- %" PRIu64 " pages to move between %" PRId64 " and %" PRId64,
+          block->name, n, last, file_size);
+
+        WT_EXT_FOREACH_FROM_OFFSET_INCL(ext, &block->live.discard, compact_start_off)
+        {
+            off = ext->off;
+            size = ext->size;
+            WT_ASSERT(session, off >= compact_start_off || off + size >= compact_start_off);
+
+            __block_compact_trim_extent(session, compact_start_off, file_size, &off, &size);
+            if (off >= compact_start_off && size <= 0)
+                break;
+
+            n = WT_EXT_SIZE_TO_LEAF_PAGES(size);
+            pages_to_move -= WT_MIN(n, pages_to_move);
+            __wt_verbose_debug3(session, WT_VERB_COMPACT,
+              "%s: estimating -- %" PRIu64 " pages on discard list between %" PRId64
+              " and %" PRId64,
+              block->name, n, off, off + size);
+        }
+        if (pages_to_move == 0)
+            break;
+
```

Comment:
In the two iteration loops above, you first estimate the number of pages to move by iterating over available extents. Then you are iterative over discarded extents in the same file area and you subtract any pages you found there from the number of pages to move calculated earlier. 

I found this confusing, because this implies that available extends may include over overlap with discarded extents. If this is the case and is the correct behaviour, please comment. Otherwise, I'd like to understand what I misunderstood (offline is okay).

Reply:
Good question! I put in the following comment—I hope it's accurate :)
```
        /*
         * Estimate how many pages we would like to move, just using the live checkpoint. The
         * checkpoint doesn't have a complete list of allocated extents, so we estimate it in two
         * phases: We first take an inverse of the "available" list, which gives us all extents that
         * are either currently allocated or are to be discarded at the next checkpoint. We do this
         * by first estimating the number of pages that can fit in the inverse of the "available"
         * list, and then we subtract the number of pages determined from the "discard" list.
         */
```